### PR TITLE
RemoteExecutor: support single file and NativeAOT

### PR DESCRIPTION
**Pull request by AustinWise**
# Problem

Currently the remote executor is used as both a library and an executable. When used as a library, the remote executor will start itself as an executable to remotely execute a method. For CoreCLR's single-file mode and NativeAOT, there is no separate file on disk for the remote executor executable, so it is not possible to start in this way.

# Solution

This PR exposes an entry point to the main method of the remote executor so that single files hosts like dotnet/runtime's
[SingleFileTestRunner](https://github.com/dotnet/runtime/blob/385e1bafaa307736541c6024ab29008f98f400f8/src/libraries/Common/tests/SingleFileTestRunner/SingleFileTestRunner.cs)
can also act as a remote remote executor host. A special argument is added to the command line so that the single file host knows whether it should execute as normal or execute as the remote executor. A copy of the `SingleFileTestRunner` is checked in, demonstrating this approach.

Note the reason this works at all with NativeAOT is the `SingleFileTestRunner` is build using TrimMode=partial. This trim mode is required for Xunit to function. I also experimented with making a [source generator](https://github.com/AustinWise/RemoteExecutorSourceGenerator/tree/SourceGenerator) version of this library that is trim compatible, but that would require users of this library to update how they define remotely-callable methods.

# Open Questions

* Currently when building as single file and NativeAOT, two executables are created: one for the `SingleFileTestRunner` and one for `Microsoft.DotNet.RemoteExecutor`. I'm not sure if this will cause problems downstream.

# Testing

* [x] The existing remote executor tests continue to pass.
* [ ] Integrate tests for single file and NativeAOT into CI?

Related bugs: #5129 dotnet/runtime#77472

# Downstream

If this is merged, dotnet/runtime#78144 will have to be cherry-picked into the commit that update the remote executor Nuget package.

### CODE REVIEW

**Comment by AustinWise** [comment]
The build analysis says "This failure is expected and is not the reason why your PR failed". Can that be trusted, or should I look into the CI failures? I don't think the Apple Simulator supports the remote executor and failing tests (System.Numerics.Vectors, which I assume is from dotnet/runtime?) don't use remote executor.

**Comment by AustinWise** [comment]
Updated to prevent infinite recursion. Also switched to using an environmental variable instead of a command line argument to trigger the remote executor.

File: src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs [hunk 1]
```diff
@@ -54,37 +57,45 @@ static RemoteExecutor()
 
             Path = typeof(RemoteExecutor).Assembly.Location;
 
-            if (IsNetCore())
+            if (string.IsNullOrEmpty(Path))
```
Comment (akoeplinger) [inline src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs:60]:
```suggestion
            if (IsSingleFile())
```

File: src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs [hunk 2]
```diff
@@ -95,6 +106,8 @@ static RemoteExecutor()
         private static bool IsNetCore() =>
             Environment.Version.Major >= 5 || RuntimeInformation.FrameworkDescription.StartsWith(".NET Core", StringComparison.OrdinalIgnoreCase);
 
+        private static bool IsSingleFile() => string.IsNullOrEmpty(typeof(RemoteExecutor).Assembly.Location);
```
Comment (akoeplinger) [inline src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs:109]:
```suggestion
        private static bool IsSingleFile() => string.IsNullOrEmpty(Path);
```

File: src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs [hunk 3]
```diff
@@ -16,6 +17,8 @@ namespace Microsoft.DotNet.RemoteExecutor
 {
     public static partial class RemoteExecutor
     {
+        internal const string REMOTE_EXECUTOR_ENVIRONMENTAL_VARIABLE = "DOTNET_REMOTE_EXECUTOR";
```
Comment (akoeplinger) [inline src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs:20]: nit: to be more consistent with `DOTNET_REMOTEEXECUTOR_SUPPORTED`
```suggestion
        internal const string REMOTE_EXECUTOR_ENVIRONMENTAL_VARIABLE = "DOTNET_REMOTEEXECUTOR";
```

**Comment by akoeplinger** [comment]
It looks good to me but I wonder if we really need the SingleFileTestRunner here. We'll notice if something breaks pretty quickly in dotnet/runtime so the code duplication doesn't seem that valuable to me. If we want to have it here we could put the SingleFileTestRunner into a package similar to [Microsoft.DotNet.XUnitConsoleRunner](https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.XUnitConsoleRunner) and remove it from dotnet/runtime.

**Comment by AustinWise** [comment]
> It looks good to me but I wonder if we really need the SingleFileTestRunner here. We'll notice if something breaks pretty quickly in dotnet/runtime so the code duplication doesn't seem that valuable to me. If we want to have it here we could put the SingleFileTestRunner into a package similar to [Microsoft.DotNet.XUnitConsoleRunner](https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.XUnitConsoleRunner) and remove it from dotnet/runtime.

Sounds good, I've removed the SingleFileTestRunner, revered the warning suppressions required to make the SingleFileTestRunner build, and adopted your other suggestions.

**Comment by agocke** [comment]
I'll try to make some time to review this. Overall, I think the long term direction we should go in with Native AOT is to use the XUnitWrapper source generator from runtime -- but that needs to be changed to xunit compatible. I have something very simple that could probably be folded in, but haven't had the time: https://github.com/agocke/xunitgen.

I'm not as familiar with the remote executor so I'll look at that code in particular.

**Comment by AustinWise** [comment]
> I'll try to make some time to review this. Overall, I think the long term direction we should go in with Native AOT is to use the XUnitWrapper source generator from runtime -- but that needs to be changed to xunit compatible. I have something very simple that could probably be folded in, but haven't had the time: https://github.com/agocke/xunitgen.
> 
> I'm not as familiar with the remote executor so I'll look at that code in particular.

Looking at the XUnitWrapper source generator, it generates the application entry point:

https://github.com/AustinWise/runtime/blob/cff9fd93a561f195a5a57d7e3935fbb0f253b6c8/src/tests/Common/XUnitWrapperGenerator/XUnitWrapperGenerator.cs#L207-L213

So it should be compatible with the approach in this PR, as it could generate code to vector into the remote executor.

For reference, here is what the changes to the SingleFileTestRunner look like to support the remote executor:

https://github.com/dotnet/runtime/pull/78144

**Comment by MichalStrehovsky** [comment]
For comparison, this is the RemoteExecutor implementation we used in .NET Native timeframe: #4142 (this is where it got ripped out)

(For .NET Native we didn't have a SingleFileRunner because we just used the runner from here, but that's orthogonal - we should be using a source generated runner instead at some point.)

**Pull request closed without merging**